scheduler/cloudformation/cloudformation.go

```diff
@@ -280,13 +288,106 @@ func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App,
 	return nil
 }
 
-// updateStack performs a stack update, but waits for the stack to enter a
-// stable state before starting.
+// updateStack enqueues the stack update, and waits up to 10 seconds for the
+// stack updated to be submitted.
+func (s *Scheduler) updateStack(input *cloudformation.UpdateStackInput, wait bool) error {
+	done := make(chan error)
+	locked := make(chan struct{})
+	errCh := make(chan error)
+
+	go func() {
+		errCh <- s.enqueueStackUpdate(input, locked, done)
+	}()
+
+	var err error
+	select {
+	case <-time.After(lockWait):
+		// FIXME: At this point, we don't want to affect UX by waiting
+		// around, so we return. But, if the stack update times out, or
+		// there's an error, that information is essentially silenced.
+		return nil
+	case err = <-errCh:
+	case <-locked:
+		// if a lock is obtained within the time frame, we might as well
+		// just wait for the update to get submitted.
+		err = <-errCh
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// If wait is set, we'll wait for the stack update to fully complete.
+	if wait {
+		return <-done
+	}
+
+	return nil
+}
+
+// enqueueStackUpdate enqueues the stack update. This function returns as soon
+// as the stack update has been submitted.
+//
+// * If there are no updates currently in progress, the stack will be updated .
+// * If there is a currently active stack update, the update will be queued behind it.
 //
-// TODO: Timeout?
-func (s *Scheduler) updateStack(stack *cloudformation.Stack, input *cloudformation.UpdateStackInput, wait bool) (*cloudformation.UpdateStackOutput, error) {
-	status := *stack.StackStatus
-	stackName := input.StackName
+// An error will be sent on the `done` channel when the stack update completes.
+func (s *Scheduler) enqueueStackUpdate(input *cloudformation.UpdateStackInput, locked chan struct{}, done chan error) error {
```

Comment:
Maybe we can generalize the locking a bit so that it can be reused for other stack operations (e.g. destroy).